File: leetcode/329. Longest Increasing Path in a Matrix.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int m = matrix.size();
        if(m == 0){
            return 0;
        }
        int n = matrix[0].size();
        vector<vector<int>> distances(matrix);
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                distances[i][j] = -1;
            }
        }
        int res = 0;
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                res = max(res, helper(i, j, INT_MIN, distances, matrix));
            }
        }
        return res;
        
    }
private:
    int helper(int x, int y, int last, vector<vector<int>>& distances, const vector<vector<int>>& matrix){
        int m = matrix.size();
        int n = matrix[0].size();
        if(x < 0 || y < 0 || x >= m || y >= n || matrix[x][y] <= last){
            return 0;
        }
        if(distances[x][y] != -1){
            return distances[x][y];
        }
        
        int left = helper(x - 1, y, matrix[x][y], distances, matrix);
        int right = helper(x + 1, y, matrix[x][y], distances, matrix);
        int up = helper(x, y - 1, matrix[x][y], distances, matrix);
        int down = helper(x, y + 1, matrix[x][y], distances, matrix);
    
        distances[x][y] = 1 + max({left, right, up, down});
        return distances[x][y];
    }
};

File: leetcode/329. Longest Increasing Path in a Matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        dirs = ((1,0),(-1,0),(0,1),(0,-1))
        m = len(matrix)
        n = len(matrix[0])
        path = [[-1 for _ in range(n)] for _ in range(m)]

        def traverse(x, y):
            if path[x][y] != -1:
                return path[x][y]
            path[x][y] = 1
            for x0, y0 in dirs:
                x1 = x0+x
                y1 = y0+y
                if x1 < 0 or x1 > m-1 or y1 < 0 or y1 > n-1:
                    continue
                if matrix[x1][y1] >= matrix[x][y]:
                    continue
                path[x][y] = max(path[x][y], traverse(x1, y1)+1)
            return path[x][y]
        
        for i in range(m):
            for j in range(n):
                traverse(i, j)

        return max([max(row) for row in path])


class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        
        dp = [[-1 for _ in range(n)] for _ in range(m)]
        directions = ((1, 0), (-1, 0), (0, 1), (0, -1))
        
        def get_path(x, y, last):
            if x < 0 or x > m - 1 or y < 0 or y > n - 1:
                return 0
            if last is not None and last >= matrix[x][y]:
                return 0
            
            if dp[x][y] != -1:
                return dp[x][y]
            
            for (x0, y0) in directions:
                dp[x][y] = max(dp[x][y], get_path(x+x0, y+y0, matrix[x][y]) + 1)
            return dp[x][y]
        
        res = 0
        for i in range(m):
            for j in range(n):
                res = max(res, get_path(i, j, None))
        return res
